src/NServiceBus.AcceptanceTests/Recoverability/Retries/When_message_goes_through_all_retries.cs

```diff
@@ -0,0 +1,176 @@
+    using EndpointTemplates;
+    using Features;
+    using NServiceBus.Config;
+    using NServiceBus.Pipeline;
+    using NUnit.Framework;
+    using ScenarioDescriptors;
+
+    public class When_message_goes_through_all_retries : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_not_send_outgoing_messages_using_batched_dispatch()
+        {
+            await Scenario.Define<Context>(c => c.Id = Guid.NewGuid())
+                .WithEndpoint<Endpoint>(b => b.DoNotFailOnErrorMessages()
+                    .When((session, context) => session.SendLocal(new IncomingMessage
+                    {
+                        Id = context.Id,
+                        UseImmediateDispatch = false
+                    }))
+                )
+                .WithEndpoint<ErrorSpy>()
+                .Done(c => c.MessageMovedToErrorQueue)
+                .Repeat(r => r.For<AllNativeMultiQueueTransactionTransports>())
+                .Should(c => Assert.IsFalse(c.OutgoingMessageReceived, "Outgoing message was unexpectedly dispatched"))
+                .Should(c => Assert.AreEqual(4, c.NumberOfProcessingAttempts))
+                .Run();
+        }
+
+        [Test]
+        public async Task Should_not_send_outgoing_messages_using_immediate_dispatch()
+        {
+            await Scenario.Define<Context>(c => c.Id = Guid.NewGuid())
+                .WithEndpoint<Endpoint>(b => b.DoNotFailOnErrorMessages()
+                    .When((session, context) => session.SendLocal(new IncomingMessage
+                    {
+                        Id = context.Id,
+                        UseImmediateDispatch = true
+                    }))
+                )
+                .WithEndpoint<ErrorSpy>()
+                .Done(c => c.MessageMovedToErrorQueue)
+                .Repeat(r => r.For<AllNativeMultiQueueTransactionTransports>())
+                .Should(c => Assert.IsFalse(c.OutgoingMessageReceived, "Outgoing message was unexpectedly dispatched"))
+                .Should(c => Assert.AreEqual(4, c.NumberOfProcessingAttempts))
+                .Run();
+        }
+
+        const string ErrorSpyQueueName = "error_spy_queue";
+
+        class Context : ScenarioContext
+        {
+            public Guid Id { get; set; }
+            public bool MessageMovedToErrorQueue { get; set; }
+            public bool OutgoingMessageReceived { get; set; }
+            public int NumberOfProcessingAttempts { get; set; }
+        }
+
+        class Endpoint : EndpointConfigurationBuilder
+        {
+            public Endpoint()
+            {
+                EndpointSetup<DefaultServer>((config, context) =>
+                {
+                    config.DisableFeature<FirstLevelRetries>();
+                    config.EnableFeature<SecondLevelRetries>();
+                    config.EnableFeature<TimeoutManager>();
+                    config.Pipeline.Register(new RegisterThrowingBehavior("SecondLevelRetries"));
```

Comment:
yes you're right since you're using InsertAfter. Mixed it up with InsertBefore. Not a big fan of the parameter though.